File: jwt/python/jwt/sign.py
from typing import Any, Dict

from pydantic import BaseModel
from util.date import get_days_in_miliseconds, now
from util.parser import generate_hmac_signature, get_base64_str


class Options(BaseModel):
    data: Dict[str, Any]
    exp_days: int
    secret: str


def sign(options: Options) -> str:
    n = now()
    header = {
        "alg": "HS256",
        "typ": "JWT",
    }
    payload = {
        **options.data,
        "iat": n,
        "exp": n + get_days_in_miliseconds(options.exp_days),
    }

    base64_header = get_base64_str(header)
    base64_payload = get_base64_str(payload)

    signature = generate_hmac_signature(
        secret=options.secret, header=base64_header, payload=base64_payload
    )

    return f"{base64_header}.{base64_payload}.{signature}"

File: jwt/ts/src/jwt/sign.ts
import { getDaysInMiliseconds } from "../util/date";
import { generateHmacSignature, getBase64String } from "../util/parser";

type Options = {
	data: Record<string, any>;
	expDays: number;
	secret: string;
};

const sign = ({ data, expDays, secret }: Options): string => {
	const now = Date.now();

	const header = {
		alg: "HS256",
		typ: "JWT",
	};

	const payload = {
		...data,
		iat: now,
		exp: now + getDaysInMiliseconds(expDays),
	};

	const base64Header = getBase64String(header);
	const base64Payload = getBase64String(payload);

	const signature = generateHmacSignature({
		secret: secret,
		header: base64Header,
		payload: base64Payload,
	});

	return `${base64Header}.${base64Payload}.${signature}`;
};

export { sign };
